catalog/app/containers/Search/model.ts

```diff
@@ -698,50 +706,328 @@ export function useNextPagePackagesQuery(after: string) {
+  'app/containers/Search:AvailableFiltersState' as const,
+  {
+    Loading: () => {},
+    Empty: () => {},
+    Ready: (x: {
+      filtering: FacetsFilteringStateInstance
+      facets: {
+        available: readonly PackageUserMetaFacet[]
+        visible: FacetTree
+        hidden: FacetTree
+      }
+      fetching: boolean
+    }) => x,
+  },
+)
+
+export type AvailableFiltersStateInstance = tagged.InstanceOf<
+  typeof AvailableFiltersState
+>
+
+type RenderFn<T> = (arg: T) => JSX.Element | null
+
+interface RenderProps<T> {
+  children: RenderFn<T>
+}
+
+export function AvailablePackagesMetaFilters({
+  children,
+}: RenderProps<AvailableFiltersStateInstance>) {
+  const model = useSearchUIModelContext(ResultType.QuiltPackage)
 
-  const query = useMetaFacetsQuery(model.state)
+  const filter = PackagesSearchFilterIO.toGQL(model.state.filter)
 
-  const all = GQL.fold(query, {
+  const query = GQL.useQuery(META_FACETS_QUERY, {
+    searchString: model.state.searchString,
+    buckets: model.state.buckets,
+    filter,
+  })
+
+  return GQL.fold(query, {
     data: ({ searchPackages: r }) => {
       switch (r.__typename) {
         case 'EmptySearchResultSet':
-          return NO_FACETS
         case 'InvalidInput':
-          return NO_FACETS
+          return children(AvailableFiltersState.Empty())
         case 'PackagesSearchResultSet':
-          return r.stats.userMeta
+          return React.createElement(AvailablePackagesMetaFiltersReady, {
+            facets: r.stats.userMeta,
+            truncated: r.stats.userMetaTruncated,
+            children,
+          })
         default:
           assertNever(r)
       }
     },
-    fetching: () => NO_FACETS,
-    error: () => NO_FACETS,
+    fetching: () => children(AvailableFiltersState.Loading()),
+    error: () => children(AvailableFiltersState.Empty()),
   })
-  const activated = model.state.userMetaFilters.filters
-  const activatedPaths = React.useMemo(() => Array.from(activated.keys()), [activated])
+}
+
+function AvailablePackagesMetaFiltersReady({
+  facets,
+  truncated,
+  children,
+}: RenderProps<AvailableFiltersStateInstance> & {
+  facets: readonly PackageUserMetaFacet[]
+  truncated: boolean
+}) {
+  const { filters } = useSearchUIModelContext(ResultType.QuiltPackage).state
+    .userMetaFilters
 
   const available = React.useMemo(
-    () => all.filter((f) => !activatedPaths.includes(f.path)),
-    [all, activatedPaths],
+    () => facets.filter((f) => !filters.has(f.path)),
+    [facets, filters],
   )
 
-  return { all, activated, activatedPaths, available, fetching: query.fetching }
+  // server-side filtering required
+  if (truncated) {
+    return React.createElement(AvailablePackagesMetaFiltersServerFilter, {
+      available,
+      children,
+    })
+  }
+
+  // client-side filtering required
+  if (available.length >= FACETS_FILTER_THRESHOLD) {
+    return React.createElement(AvailablePackagesMetaFiltersClientFilter, {
```

Comment:
> Do you use React.createElement ... because model.ts has .ts extension, not .tsx?

yeah mostly

> Probably, another reason is to highlight the fact, these components don't contain any UI, and they need just to render hooks?

this, too, but mostly as a coincidence